onmt/train_single.py

```diff
@@ -106,7 +107,13 @@ def main(opt, device_id):
     # (All datasets have the same data_type).
     # this should be refactored out of existence reasonably soon
     first_dataset = torch.load(glob.glob(opt.data + '.train*.pt')[0])
```

Comment:
why not use the valid set for this ?

Reply:
I suppose that would make sense as well. Using an arbitrary train shard is the precedent. I'm not inclined to change it around because my feeling is that it shouldn't be necessary to inspect a dataset like this at all (hence the comment about refactoring it out of existence).

Reply:
As a matter of fact I realized this is crucial point in my use case.
When working on 6 GPU, with a shard of 1Mio examples, (825MB) this stupid check will load 6 times in memory.
My machine has 64GB and it swaps for ever. When changing to valid-set, no problem at all.

Reply:
This point is well taken. When I can next take a look, first I'll see if the check can be removed completely. This seems likely to me because the ways the `data_type` is actually used are pretty marginal.

Until then, checking validation instead of training is a good stopgap. Also, I don't understand why it's necessary to do this on all GPUs, but that's an issue to revisit only if it isn't possible to remove this dataset loading for unforeseen reasons.

Loading a training shard to check the data type is not new -- it has been present in the repository for several months. Have you previously trained with this dataset without problems?

Reply:
Yes plenty of runs were fine, but probably at the edge. When I ran one this afternoon, it was stuck not seeing why. I realized it worked fine with 5 GPU, not 6, then I saw the swapping for ever.
Moreovern I don't see any recent change that may have changed the memory behavior before the actual training starts.

Reply:
Exactly -- I would be quite surprised if this were a new problem. I think it just hadn't been noticed before.

Reply:
I was able to remove the `first_dataset` bit completely. The trainer has access to `model_opt.model_type`, which has the same values as `data_type`. This makes loading a dataset to find the data type unnecessary.

I was also able to remove the `.data_type` attribute from `Trainer`. The only thing `data_type` was used for was `make_features`, a function which I hope to be able to get rid of at a later date. `make_features` is essentially for post-processing the batch data to create the actual input to the network. It's quite messy that currently it needs to be present at training, validation, and translation time.

Reply:
ok great.
in the meantime (before removing make_features) can you refactor a bit a make it the common function that returns src, src_length = make_features() ?
you have 3 times the same code (twice in trainer, once in run_encoder)

Reply:
Done.